File: forester/ruby/evoruby/lib/evo/table/basic_table.rb
# = lib/evo/table/basic_table.rb - BasicTable class
#
# Copyright::  Copyright (C) 2006-2007 Christian M. Zmasek
# License::    GNU Lesser General Public License (LGPL)'s
#
# $Id: basic_table.rb,v 1.3 2007/09/28 03:12:10 cmzmasek Exp $
#
# last modified: 05/16/2007

#require 'lib/evo/util/constants'

module Evoruby

  class BasicTable

        def initialize()
            @rows = Hash.new
            @max_row = 0
            @max_col = 0
        end

        # raises ArgumentError
        def set_value( row, col, value )
            if ( ( row < 0 ) || ( col < 0 ) )
                raise( ArgumentError, "attempt to use negative values for row or column" )
            end
            if ( row > get_max_row() )
                set_max_row( row )
            end
            if ( col > get_max_col() )
                set_max_col( col )
            end
            row_map = nil
            if ( @rows.has_key?( row ) )
                row_map = @rows[ row ]
            else
                row_map = Hash.new
                @rows[ row ] = row_map
            end
            row_map[ col ] = value
        end

        # raises ArgumentError
        def get_value_as_string( row, col )
            return ( get_value( row, col ) ).to_s
        end

        # raises ArgumentError
        def get_value( row, col )
            if ( ( row > get_max_row() ) || ( row < 0 ) )
                raise( ArgumentError, "value for row (" + row.to_s +
                         ") is out of range [max row: " + get_max_row().to_s + "]" )
            elsif ( ( col > get_max_col() ) || ( row < 0 ) )
                raise( ArgumentError, "value for column (" + col.to_s +
                         ") is out of range [max column: " + get_max_col().to_s + "]" )
            end
            row_map = @rows[ row ]
            if ( ( row_map == nil ) || ( row_map.length < 1 ) )
                return nil
            end
            return row_map[ col ]
        end

        def get_max_col()
            return @max_col
        end

        def get_max_row()
            return @max_row
        end

        # raises ArgumentError
        def get_columns_as_map( key_col, value_col )
            map = Hash.new
            for row in 0 .. get_max_row
                key = get_value( row, key_col )
                value = get_value( row, value_col )
                if ( ( key != nil ) && ( value != nil ) )
                    if ( map.has_key?( key ) )
                        raise( ArgumentError, "attempt to use non-unique table value as key [" +
                                        + key + "]" )
                    end
                    map[ key ] = value
                end
            end
            return map
        end

        def to_s
            str = String.new
            for row in 0 .. get_max_row
               for col in 0 .. get_max_col
                   str << col.to_s << " "
               end
               str << LEvoruby::Constants::LINE_DELIMITER
               for col in 0 .. get_max_col
                   str << row.to_s << ": "
                   str << get_value( row, col ) << " "
               end
               str << Evoruby::Constants::LINE_DELIMITER
            end
            return str
        end


        private

        def get_row( row )
            return @rows[ row ]
        end

        def set_max_col( max_col )
            @max_col = max_col
        end

        def set_max_row( max_row )
            @max_row = max_row
        end

    end # class BasicTable

end # module Evoruby

File: forester/java/src/org/forester/util/BasicTable.java
// $Id:
// FORESTER -- software libraries and applications
// for evolutionary biology research and applications.
//
// Copyright (C) 2008-2009 Christian M. Zmasek
// Copyright (C) 2008-2009 Burnham Institute for Medical Research
// All rights reserved
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation; either
// version 2.1 of the License, or (at your option) any later version.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
//
// Contact: phylosoft @ gmail . com
// WWW: www.phylosoft.org

package org.forester.util;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class BasicTable<E> {

    private int                         _max_col;
    private int                         _max_row;
    private Map<String, Map<String, E>> _rows;

    public BasicTable() {
        init();
    }

    // Returns -1 if not found, IllegalArgumentException if not unique.
    public int findRow( final String first_col_value ) throws IllegalArgumentException {
        int result = -1;
        for( int i = 0; i < this.getNumberOfRows(); ++i ) {
            if ( getValueAsString( 0, i ).equals( first_col_value ) ) {
                if ( result >= 0 ) {
                    throw new IllegalArgumentException( "\"" + first_col_value + "\" is not unique" );
                }
                result = i;
            }
        }
        return result;
    }

    public Map<String, E> getColumnsAsMap( final int key_col, final int value_col ) throws IllegalArgumentException {
        final Map<String, E> map = new HashMap<String, E>();
        for( int row = 0; row < getNumberOfRows(); ++row ) {
            final String key = ( String ) getValue( key_col, row );
            final E value = getValue( value_col, row );
            if ( ( key != null ) && ( value != null ) ) {
                if ( map.containsKey( key ) ) {
                    throw new IllegalArgumentException( "attempt to use non-unique table value as key [" + key + "]" );
                }
                map.put( key, value );
            }
        }
        return map;
    }

    public Map<String, Double> getColumnsAsMapDouble( final int key_col, final int value_col )
            throws IllegalArgumentException, IOException {
        final Map<String, Double> map = new HashMap<String, Double>();
        for( int row = 0; row < getNumberOfRows(); ++row ) {
            final String key = ( String ) getValue( key_col, row );
            double value = 0;
            try {
                value = Double.parseDouble( getValueAsString( value_col, row ) );
            }
            catch ( final NumberFormatException e ) {
                throw new IOException( e );
            }
            if ( key != null ) {
                if ( map.containsKey( key ) ) {
                    throw new IllegalArgumentException( "attempt to use non-unique table value as key [" + key + "]" );
                }
                map.put( key, value );
            }
        }
        return map;
    }

    public int getNumberOfColumns() {
        return _max_col + 1;
    }

    public int getNumberOfRows() {
        return _max_row + 1;
    }

    public final String getRowAsString( final int row, final String separator ) {
        final StringBuilder sb = new StringBuilder();
        for( int col = 0; col < getNumberOfColumns(); ++col ) {
            sb.append( getValue( col, row ).toString() );
            if ( col < ( getNumberOfColumns() - 1 ) ) {
                sb.append( separator );
            }
        }
        return sb.toString();
    }

    public E getValue( final int col, final int row ) throws IllegalArgumentException {
        if ( ( row > ( getNumberOfRows() - 1 ) ) || ( row < 0 ) ) {
            throw new IllegalArgumentException( "value for row (" + row + ") is out of range [number of rows: "
                    + getNumberOfRows() + "]" );
        }
        else if ( ( col >= getNumberOfColumns() ) || ( row < 0 ) ) {
            throw new IllegalArgumentException( "value for column (" + col + ") is out of range [number of columns: "
                    + getNumberOfColumns() + "]" );
        }
        final Map<String, E> row_map = getRow( row );
        if ( ( row_map == null ) || ( row_map.size() < 1 ) ) {
            return null;
        }
        return row_map.get( "" + col );
    }

    public String getValueAsString( final int col, final int row ) throws IllegalArgumentException {
        if ( getValue( col, row ) != null ) {
            return getValue( col, row ).toString();
        }
        return null;
    }

    public boolean isEmpty() {
        return getNumberOfRows() <= 0;
    }

    public void setValue( final int col, final int row, final E value ) {
        if ( ( row < 0 ) || ( col < 0 ) ) {
            throw new IllegalArgumentException( "attempt to use negative values for row or column" );
        }
        if ( row > ( getNumberOfRows() - 1 ) ) {
            setMaxRow( row );
        }
        if ( col > ( getNumberOfColumns() - 1 ) ) {
            setMaxCol( col );
        }
        final String row_key = "" + row;
        Map<String, E> row_map = null;
        if ( getRows().containsKey( row_key ) ) {
            row_map = getRows().get( row_key );
        }
        else {
            row_map = new HashMap<String, E>();
            getRows().put( row_key, row_map );
        }
        row_map.put( "" + col, value );
    }

    @Override
    public String toString() {
        final StringBuilder sb = new StringBuilder();
        for( int row = 0; row < getNumberOfRows(); ++row ) {
            for( int col = 0; col < getNumberOfColumns(); ++col ) {
                sb.append( getValue( col, row ) );
                if ( col < ( getNumberOfColumns() - 1 ) ) {
                    sb.append( " " );
                }
            }
            if ( row < ( getNumberOfRows() - 1 ) ) {
                sb.append( ForesterUtil.LINE_SEPARATOR );
            }
        }
        return sb.toString();
    }

    private Map<String, E> getRow( final int row ) {
        return getRows().get( "" + row );
    }

    private Map<String, Map<String, E>> getRows() {
        return _rows;
    }

    private void init() {
        _rows = new HashMap<String, Map<String, E>>();
        setMaxCol( -1 );
        setMaxRow( -1 );
    }

    private void setMaxCol( final int max_col ) {
        _max_col = max_col;
    }

    private void setMaxRow( final int max_row ) {
        _max_row = max_row;
    }
}
